File: 17. Letter Combinations of a Phone Number/Recursive Writing/letter_combinations_of_a_phone_number.swift
let combinations: Array<String> = [" ", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"];

func word_from_odo (_ relevant_digits: String,_ digits_odo: Array<Int>) -> String{
	var nje: String = "";
	for i in 0..<digits_odo.count{
		let rindex = relevant_digits.index(relevant_digits.startIndex, offsetBy: i);
		let digit: Int = Int(String(relevant_digits[rindex]))!;
		let button: String = combinations[digit];
		let nindex = button.index(button.startIndex, offsetBy: digits_odo[i]);
		nje.append(button[nindex]);
	}
	
	return (nje);
}

func word_fill(_ relevant_digits: String,_ digits_odo: Array<Int>,_ depth: Int,_ sample_space: inout Array<String>) -> (){
	let rindex = relevant_digits.index(relevant_digits.startIndex, offsetBy: depth);
	let len_letter_options: Int = combinations[Int(String(relevant_digits[rindex]))!].count;
	let max_depth: Int = relevant_digits.count;
	var temp_digits_odo: Array<Int> = digits_odo;
	
	for i in 0..<len_letter_options{
		temp_digits_odo[depth] = i;
		if (depth < (max_depth - 1)){
			word_fill (relevant_digits, temp_digits_odo, (depth + 1), &sample_space);
		} else {
			let word: String = word_from_odo (relevant_digits, temp_digits_odo);
			sample_space.append(word);
		}
	}
	
}

func get_letter_combinations (_ digits: String) -> Array<String>{
	var nje: Array<String> = [];
	
	let len_digits: Int = digits.count;
	var odo: Array<Int> = [];
	
	var useable_digits: String = "";
	
	for i in 0..<len_digits{
		let dindex = digits.index(digits.startIndex, offsetBy: i);
		let no_of_letters = combinations[Int(String(digits[dindex]))!].count;
		
		if (no_of_letters == 0){
			continue;
		}
		
		useable_digits.append(digits[dindex]);
		
		odo.append(0);
	}
	
	//let len_useable_digits = useable_digits.count;
	word_fill (useable_digits, odo, 0, &nje);
	
	return (nje);
}

let digits: String = "68072";
//let odo: Array<Int> = Array(repeating: 1, count: digits.count);
print(get_letter_combinations(digits));

File: 17. Letter Combinations of a Phone Number/Recursive Writing/letter_combinations_of_a_phone_number.rs
const combinations: [&str; 10] = [" ", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"];

fn word_from_odo (relevant_digits: &String, digits_odo: &Vec<usize>) -> String {
	let mut nje: String = "".to_string();
	let length: usize = relevant_digits.len();
	
	for i in 0..length{
		let digit: usize = relevant_digits.chars().nth(i).unwrap().to_digit(10).unwrap() as usize;
		nje.push(combinations[digit].chars().nth(digits_odo[i]).unwrap());
	}
	
	return (nje);
}

fn word_fill (relevant_digits: &String, digits_odo: &Vec<usize>, depth: usize, mut sample_space: &mut Vec<String>) -> (){
	let len_letter_options: usize = combinations[relevant_digits.chars().nth(depth).unwrap().to_digit(10).unwrap() as usize].len();
	let max_depth: usize = relevant_digits.len();
	let mut temp_digits_odo: Vec<usize> = digits_odo.clone();
	
	for i in 0..len_letter_options{
		temp_digits_odo[depth] = i;
		if (depth < (max_depth - 1)){
			word_fill (&relevant_digits, &temp_digits_odo, (depth + 1), &mut sample_space);
		} else {
			let word: String = word_from_odo(relevant_digits, &temp_digits_odo);
			sample_space.push(word);
		}
	}
}

fn get_letter_combinations (digits: String) -> Vec<String> {
	let len_digits: usize = digits.len();
	let mut useable_digits: String = "".to_string();
	let mut odo: Vec<usize> = vec![];
	let mut nje: Vec<String> = vec![];
	
	for i in 0..len_digits{
		let len_letters: usize = combinations[digits.chars().nth(i).unwrap().to_digit(10).unwrap() as usize].len();
		if (len_letters == 0){
			continue;
		}
		
		useable_digits.push(digits.chars().nth(i).unwrap());
		odo.push(0);
	}
	
	word_fill (&useable_digits, &odo.clone(), 0, &mut nje);
	
	return (nje);
}

fn main () {
	let digits: String = "68072".to_string();
	let words: Vec<String> = get_letter_combinations(digits);
	
	println!("This is the handy work of Rust.\n\n\n");
	
	for i in words{
		println!("{}", i);
	}
	
	println!("\n\n\n");
}
